drivers/can/can_max32.c

```diff
@@ -0,0 +1,752 @@
+{
+	ARG_UNUSED(dev);
+
+	*cap = CAN_MODE_NORMAL | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY;
+
+	return 0;
+}
+
+static int can_max32_set_mode(const struct device *dev, can_mode_t mode)
+{
+	const struct max32_can_config *dev_cfg = dev->config;
+	struct max32_can_data *dev_data = dev->data;
+	int ret = 0;
+	can_mode_t cap;
+
+	can_get_capabilities(dev, &cap);
+
+	if (dev_data->common.started) {
+		return -EBUSY;
+	}
+
+	if ((mode & ~(cap)) != 0) {
+		LOG_ERR("unsupported mode: 0x%08x; Capabilities: 0x%08x", mode, cap);
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&dev_data->inst_mutex, K_FOREVER);
+
+	if (((mode & CAN_MODE_LOOPBACK) != 0) && ((mode & CAN_MODE_LISTENONLY) != 0)) {
+		LOG_DBG("Set mode CAN_MODE_LOOPBACK and CAN_MODE_LISTENONLY");
+		ret = MXC_CAN_SetMode(dev_cfg->can_id, MXC_CAN_MODE_LOOPBACK_W_TXD);
+	} else if ((mode & CAN_MODE_LOOPBACK) != 0) {
+		LOG_DBG("Set mode CAN_MODE_LOOPBACK");
+		ret = MXC_CAN_SetMode(dev_cfg->can_id, MXC_CAN_MODE_LOOPBACK);
+	} else if ((mode & CAN_MODE_LISTENONLY) != 0) {
+		LOG_DBG("Set mode CAN_MODE_LISTENONLY");
+		ret = MXC_CAN_SetMode(dev_cfg->can_id, MXC_CAN_MODE_MONITOR);
+	} else {
+		LOG_DBG("Set mode CAN_MODE_NORMAL");
+		ret = MXC_CAN_SetMode(dev_cfg->can_id, MXC_CAN_MODE_NORMAL);
+	}
+
+	if (ret == E_NO_ERROR) {
+		dev_data->common.mode = mode;
+	}
+
+	k_mutex_unlock(&dev_data->inst_mutex);
+
+	return ret == E_NO_ERROR ? 0 : -ECANCELED;
+}
+
+static int can_max32_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct max32_can_config *dev_cfg = dev->config;
+
+	*rate = MXC_CAN_GetClock(dev_cfg->can_id);
+
+	return 0;
+}
+
+static int can_max32_init_timing_struct(struct can_timing *timing, const struct device *dev)
+{
+	const struct max32_can_config *dev_cfg = dev->config;
+	int ret = 0;
+
+	ret = can_calc_timing(dev, timing, dev_cfg->common.bitrate,
+				dev_cfg->common.sample_point);
+	if (ret < 0) {
+		LOG_ERR("can_calc_timing error sample_point: %d!",
+			dev_cfg->common.sample_point);
+	}
+	LOG_DBG("Bitrate: %d", dev_cfg->common.bitrate);
+	LOG_DBG("Presc: %d, PG1: %d, PG2: %d", timing->prescaler,
+		timing->prop_seg + timing->phase_seg1, timing->phase_seg2);
+
+	return ret;
+}
+
+static int can_max32_set_timing(const struct device *dev, const struct can_timing *timing)
+{
+	const struct max32_can_config *dev_cfg = dev->config;
+	struct max32_can_data *dev_data = dev->data;
+	mxc_can_regs_t *can = dev_cfg->can;
+	uint32_t nbt_reg = 0;
+
+	if (!timing) {
+		LOG_ERR("timing structure is null");
+		return -EINVAL;
+	}
+
+	if (dev_data->common.started) {
+		return -EBUSY;
+	}
+
+	k_mutex_lock(&dev_data->inst_mutex, K_FOREVER);
+
+	can->mode |= MXC_F_CAN_MODE_RST;
+
+	nbt_reg |= FIELD_PREP(MXC_F_CAN_NBT_NBRP, timing->prescaler - 1);
+	nbt_reg |= FIELD_PREP(MXC_F_CAN_NBT_NSEG1, timing->prop_seg + timing->phase_seg1 - 1);
+	nbt_reg |= FIELD_PREP(MXC_F_CAN_NBT_NSEG2, timing->phase_seg2 - 1);
+	nbt_reg |= FIELD_PREP(MXC_F_CAN_NBT_NSJW, timing->sjw - 1);
+
+	can->nbt |= nbt_reg;
```

Comment:
Shouldn't this be a regular `=` instead of `|=`? Otherwise existing bits are not cleared.